File: src/dnn/asyncarray.cpp
/*
    This file is originally from gocv project
    Licensed: Apache 2.0 license. Copyright (c) 2017-2021 The Hybrid Group.

    Modified by Rainyl.
    Licensed: Apache 2.0 license. Copyright (c) 2024 Rainyl.
*/

#include "asyncarray.h"
#include <string.h>

// AsyncArray_New creates a new empty AsyncArray
CvStatus *AsyncArray_New(AsyncArray *rval)
{
  BEGIN_WRAP
  *rval = {new cv::AsyncArray()};
  END_WRAP
}

// AsyncArray_Close deletes an existing AsyncArray
void AsyncArray_Close(AsyncArrayPtr a) { CVD_FREE(a); }

CvStatus *AsyncArray_Get(AsyncArray async_out, Mat out)
{
  BEGIN_WRAP
  async_out.ptr->get(*out.ptr);
  END_WRAP
}

CvStatus *Net_forwardAsync(Net net, const char *outputName, AsyncArray *rval)
{
  BEGIN_WRAP
  auto arr = net.ptr->forwardAsync();
  *rval = {&arr};
  END_WRAP
}

File: lib/src/dnn/asyncarray.dart
library cv.dnn;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../g/dnn.g.dart' as cdnn;

class AsyncArray extends CvStruct<cdnn.AsyncArray> {
  AsyncArray._(cdnn.AsyncArrayPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }

  factory AsyncArray.fromPointer(cdnn.AsyncArrayPtr ptr, [bool attach = true]) => AsyncArray._(ptr, attach);

  factory AsyncArray.empty() {
    final p = calloc<cdnn.AsyncArray>();
    cvRun(() => cdnn.AsyncArray_New(p));
    final arr = AsyncArray._(p);
    return arr;
  }

  static final finalizer = OcvFinalizer<cdnn.AsyncArrayPtr>(cdnn.addresses.AsyncArray_Close);

  void dispose() {
    finalizer.detach(this);
    cdnn.AsyncArray_Close(ptr);
  }

  Mat get() {
    final dst = Mat.empty();
    cvRun(() => cdnn.AsyncArray_Get(ref, dst.ref));
    return dst;
  }

  @override
  cdnn.AsyncArray get ref => ptr.ref;
}
